internal_transfers/actions/tests/database.spec.ts

```diff
@@ -151,10 +154,128 @@ describe("test database insertion methods", () => {
     ];
     expect(results).toStrictEqual(expected);
 
-    // Idempotency
-    await insertSettlementSimulations(db, dummySimData);
-    results = await db.query(sql`SELECT * from settlement_simulations;`);
-    expect(results).toStrictEqual(expected);
+    await expect(insertSettlementSimulations(db, dummySimData)).rejects.toThrow(
+      'duplicate key value violates unique constraint "settlement_simulations_pkey"'
+    );
+  });
+  test("insertPipelineResults", async () => {
```

Comment:
This seems like a test with many subtests in it. In python I would expect to separate these not sure how it works here. Just a little bit hard to follow at times.

Reply:
I can try to clean it up, but I do believe it all falls into one test that is to demonstrate there is NO possible way to insert the same data twice. The function calls insert on 3 tables (synchronously).

then we 

1. write to it once (success),
2. try to write again with same stuff (fail),
3. try to write where table 1 insert would have passed (fail)
4. try to write where table 1 & 2 insert would have passed (fail)
5. write where table 1, 2, 3 insert would have passed (success)


Perhaps step 5 is overkill because we know we can write to the table.

I could also try to make the test more legible instead of randomly altering fields in the test objects.

Reply:
I was also confused by the test and had to look the summary here.
My understanding now is that we want what we insert on each of the three tables to be distinct, otherwise the transaction should fail.
Imho it would be helpful to have three tests:
1. Just insert and it works.
2. Inserting twice the same fails.
3. Complex logic (a combination of your steps 3, 4, and maybe 5): each single intermediate step cannot repeat. You can also wrap everything in a `describe` and have separate tests for each intermediate step, but this might be too much boilerplate.

Reply:
I split it up at the expense of quite a diff (we had to wrap everything in another describe block because we need to close the db connection and the db is defined globally outside anyway - I think it looks much cleaner now. Please let me know.

Reply:
glad you like them!